geoopt/manifolds/lorentz/math.py

```diff
@@ -0,0 +1,607 @@
+    d = zp.size(dim) - 1
+    zp = th.cat(
+        (zp.narrow(dim, 0, 1) * th.sqrt(k), zp.narrow(dim, 1, d) * 0.0), dim=dim
+    )
+
+    dist_ = dist(zp, y, k=k, dim=dim, keepdim=True)
+    nomin = y + 1.0 / k * inner(zp, y, keepdim=True) * zp
+    denom = norm(nomin, keepdim=True)
+    return dist_ * nomin / denom
+
+
+def egrad2rgrad(x, grad, *, k=1.0, dim=-1):
+    r"""
+    Translate Euclidean gradient to Riemannian gradient on tangent space of :math:`x`.
+
+    .. math::
+
+        \Pi_{\mathbb{R}^{d+1} \rightarrow \mathcal{T}_{\mathbf{x}} \mathbb{H}^{d, k}(\mathbf{v})}:=\mathbf{v}+\langle\mathbf{x}, \mathbf{v}\rangle_{\mathcal{L}} \frac{\mathbf{x}}{k}
+
+    Parameters
+    ----------
+    x : tensor
+        point on the Hyperboloid
+    grad : tensor
+        Euclidean gradient for :math:`x`
+    k : float|tensor
+        manifold negative curvature
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+        Riemannian gradient :math:`u\in `
+    """
+    return _egrad2rgrad(x, grad, k=k, dim=dim)
+
+
+def _egrad2rgrad(x, grad, k, dim: int = -1):
+    grad = grad.addcmul(inner(x, grad, dim=dim, keepdim=True).expand_as(x), x / k)
+    return grad
+
+
+def parallel_transport(x, y, v, *, k=1.0, dim=-1):
+    r"""
+    Perform parallel transport on the Hyperboloid.
+
+    Parameters
+    ----------
+    x : tensor
+        starting point
+    y : tensor
+        end point
+    v : tensor
+        tangent vector to be transported
+    c : float|tensor
+        manifold negative curvature
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+        transported vector
+    """
+    return _parallel_transport(x, y, v, k=k, dim=dim)
+
+
+def _parallel_transport(x, y, v, k, dim: int = -1):
+    nom = inner(logmap(x, y, k=k, dim=dim), v, keepdim=True)
+    denom = dist(x, y, k=k, dim=dim, keepdim=True) ** 2
+    p = v - nom / denom * (logmap(x, y, k=k, dim=dim) + logmap(y, x, k=k, dim=dim))
+    return p
+
+
+def parallel_transport0(y, v, *, k=1.0, dim=-1):
+    r"""
+    Perform parallel transport from zero point.
+
+    Parameters
+    ----------
+    y : tensor
+        end point
+    v : tensor
+        tangent vector to be transported
+    c : float|tensor
+        manifold negative curvature
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+        transported vector
+    """
+    return _parallel_transport0(y, v, k=k, dim=dim)
+
+
+def _parallel_transport0(y, v, k, dim: int = -1):
+    zp = th.ones_like(y)
+    d = zp.size(dim) - 1
+    zp = th.cat(
+        (zp.narrow(dim, 0, 1) * th.sqrt(k), zp.narrow(dim, 1, d) * 0.0), dim=dim
+    )
+
+    nom = inner(logmap(zp, y, k=k, dim=dim), v, keepdim=True)
+    denom = dist(zp, y, k=k, dim=dim, keepdim=True) ** 2
+    p = v - nom / denom * (logmap(zp, y, k=k, dim=dim) + logmap(y, zp, k=k, dim=dim))
+    return p
+
+
+def geodesic_unit(t, x, u, k=1.0):
```

Comment:
No `dim` argument, k should also be keyword only. As for kw only, that is the only point to have a private implementation btw. Torch script doesn't support advanced signatures.